engine/src/test/scala/cromwell/webservice/MetadataBuilderActorSpec.scala

```diff
@@ -0,0 +1,242 @@
+        |      }, {
+        |        "attempt": 2,
+        |        "NOT_CHECKED": "NOT_CHECKED",
+        |        "shardIndex": 1
+        |      }, {
+        |        "attempt": 3,
+        |        "NOT_CHECKED": "NOT_CHECKED",
+        |        "shardIndex": 1
+        |      }],
+        |      "callA": [{
+        |        "attempt": 1,
+        |        "NOT_CHECKED": "NOT_CHECKED",
+        |        "shardIndex": -1
+        |      }]
+        |    },
+        |    "NOT_CHECKED": "NOT_CHECKED"
+        |  },
+        |  "$workflowB": {
+        |    "calls": {
+        |      "callA": [{
+        |        "attempt": 1,
+        |        "NOT_CHECKED": "NOT_CHECKED",
+        |        "shardIndex": -1
+        |      }]
+        |    },
+        |    "NOT_CHECKED": "NOT_CHECKED"
+        |  }
+        |}""".stripMargin
+
+    val mdQuery = MetadataQuery(None, None, None)
+    val queryAction = GetMetadataQueryAction(mdQuery)
+    assertMetadataResponse(queryAction, mdQuery, workflowAEvents ++ workflowBEvents, expectedRes)
+  }
+
+  type EventBuilder = (String, String, DateTime)
+
+  def assertMetadataKeyStructure(eventList: List[EventBuilder], expectedJson: String) = {
+    val workflow = WorkflowId.randomId()
+
+    def makeEvent(workflow: WorkflowId)(key: String, value: String, timestamp: DateTime) = {
+      MetadataEvent(MetadataKey(workflow, None, key), MetadataValue(value), timestamp)
+    }
+
+    val events = eventList map Function.tupled(makeEvent(workflow))
+    val expectedRes = s"""{ "calls": {}, $expectedJson } """
+
+    val mdQuery = MetadataQuery(Option(workflow), None, None)
+    val queryAction = GetAllMetadataAction(workflow)
+    assertMetadataResponse(queryAction, mdQuery, events, expectedRes)
+  }
+
+  it should "keep event with later timestamp for the same key in metadata" in {
+    val eventBuilderList = List(
+      ("a", "aLater", DateTime.parse("2000-01-02")),
+      ("a", "a", DateTime.parse("2000-01-01"))
+    )
+    val expectedRes =
+      """"a": "aLater"""".stripMargin
+
+    assertMetadataKeyStructure(eventBuilderList, expectedRes)
+  }
+
+  it should "build JSON object structure from dotted key syntax" in {
+    val eventBuilderList = List(
+      ("a.b.c", "abc", DateTime.now),
+      ("b.a", "ba", DateTime.now),
+      ("c", "c", DateTime.now)
+    )
+
+    val expectedRes =
+      """"a": {
+        |    "b": {
+        |      "c": "abc"
+        |    }
+        |  },
+        |  "b": {
+        |    "a": "ba"
+        |  },
+        |  "c": "c"""".stripMargin
+
+    assertMetadataKeyStructure(eventBuilderList, expectedRes)
+  }
+
+
+  it should "build lexigraphically sorted JSON list structure from dotted key syntax" in {
+    val eventBuilderList = List(
+      ("l1[2]", "l12", DateTime.now),
+      ("l1[k]", "l1k", DateTime.now),
+      ("l1[3]", "l13", DateTime.now),
+      ("l1[i]", "l1i", DateTime.now),
+      ("l1[10]", "l110", DateTime.now),
+      ("l1[j]", "l1j", DateTime.now)
```

Comment:
It'd be great to have an in-list update test here too - e.g.
```
("x[1]", "a", _)
("x[2]", "a", _)
("x[3]", "a", _)

("x[2]", "b", _)
("x[3]", "b", _)

("x[3]", "c", _)
```

should become:
```
{
  "x": [ "a", "b", "c" ]
}
```